File: data/maps/mg.py
# 
# How to:
# 1. DefineEncounter, then define Enemy's and define Encounter's
# 2. Define Group's where each group name is a DefineEncounter name
# 
# Note:
# An DefineEncounter can have at most 32 Encounters. Subject to change.
# 
# Enemy Moves:
# - down
# - downleft
# - downright
# - downsin
# - downcos
# - downstop,
# - horzright
# - horzleft
# - horzrightstop
# - horzleftstop
#
# Bullet Types:
# - straight
# - left
# - right
# - spin
# - flower
# - sin
# - sin_wide
# - big
# - beam
#

import sys

header = None
header_exists = None
enemies = None
encounters = None
groups = None

def Reset():
    global header, header_exists, enemies, encounters, groups
    header = []
    header_exists = False
    enemies = ["$enemies"]
    encounters = ["$encounters"]
    groups = ["$sets"]
Reset()

def Header(**kwargs):
    global header_exists

    if header_exists:
        print("ERROR: Only _one_ header may exist per file. Found a second header.")
        exit(1)
    header_exists = True

    header.append("**")
    for key, value in kwargs.items():
        header.append("%s %s" % (key, value))
    header.append("**")

class Enemy:
    def __init__(self, name, shoot="straight", move="down", hp=5, speed=1, meta=0, level=0, png=None):
        self.name = name
        self.shoot = "bullet_enemy_" + shoot
        self.move = move
        self.hp = hp
        self.speed = speed
        self.meta = meta
        self.level = level
        self.png = png
        enemies.append(self)

    def __repr__(self):
        return """\
(
name %s
shoot %s
move %s
hp %s
speed %s
meta %s
level %s%s
)""" % (self.name, self.shoot, self.move, self.hp, self.speed, self.meta, self.level, "" if self.png is None else "\n    png %s" % self.png)
    def __str__(self):
        return self.__repr__()

class Encounter:
    def __init__(self, name, time, x, y):
        self.name = name
        self.time = time
        self.x = x
        self.y = y
        encounters.append(self)
    def __repr__(self):
        return """\
(
name %s
time %s
x %s
y %s
)""" % (self.name, self.time, self.x, self.y)
    def __str__(self):
        return self.__repr__()

def DefineEncounter(name):
    encounters.append('[' + name + ']')

class Group:
    def __init__(self, name):
        self.name = name
        groups.append(self)
    def __repr__(self):
        return self.name
    def __str__(self):
        return self.name

def Produce():
    if len(sys.argv) > 1:
        filename = sys.argv[1]
    else:
        filename = None

    fp = sys.stdout
    if filename:
        fp = open(filename, "w")

    if header_exists:
        print("\n".join(str(b) for b in header), file=fp)

    print("\n".join(str(b) for b in enemies), file=fp)
    print("\n".join(str(b) for b in encounters), file=fp)
    print("\n".join(str(b) for b in groups), file=fp)
    
    if filename:
        fp.close()

File: data/maps/mg.lua
--[[
How to:
1. DefineEncounter, then define Enemy's and define Encounter's
2. Define Group's where each group name is a DefineEncounter name

Note:
An DefineEncounter can have at most 32 Encounters. Subject to change.

Enemy Moves:
- down
- downleft
- downright
- downsin
- downcos
- downstop,
- horzright
- horzleft
- horzrightstop
- horzleftstop

Bullet Types:
- straight
- left
- right
- spin
- flower
- sin
- sin_wide
- big
- beam
--]]

local M = {}

function M.New()
	local N = {
		header = {},
		enemies = {'$enemies'},
		encounters = {'$encounters'},
		groups = {'$sets'},
	}
	return N
end

function M.Header(self, atmosphere)
	table.insert(self.header,'**')
	table.insert(self.header, string.format("%s %s", "atmosphere", atmosphere))
	table.insert(self.header, '**')
end

function M.Enemy(self, t)
	setmetatable(t, {__index={
		name=nil,
		shoot='straight',
		move='down',
		hp=5,
		speed=1,
		meta=0,
		level=0,
		png=nil,
	}})
	local name = t[1]
	local shoot = 'bullet_enemy_' .. (t[2] or t.shoot)
	local move = t[3] or t.move
	local hp = t[4] or t.hp
	local speed = t[5] or t.speed
	local meta = t[6] or t.meta
	local level = t[7] or t.level
	local png = t[8] or t.png
	local s = string.format([[
(
name %s
shoot %s
move %s
hp %d
speed %d
meta %f
level %d%s
)]], name, shoot, move, hp, speed, meta,
			level, png and string.format("\npng %s", png) or "")
	table.insert(self.enemies, s)
end

function M.DefineEncounter(self, name)
	table.insert(self.encounters, string.format('[%s]', name))
end

function M.Encounter(self, name, time, x, y)
	local s = string.format([[
(
name %s
time %f
x %d
y %d
)]], name, time, x, y)
	table.insert(self.encounters, s)
end

function M.Group(self, name)
	table.insert(self.groups, name)
end

function M.Generate(self)
	all = {self.header, self.enemies, self.encounters, self.groups}
	for i, t in ipairs(all) do
		for j, v in ipairs(t) do
			print(v)
		end
	end
end

return M
